### PR TITLE
jni: add support for on-heap buffers

### PR DESCRIPTION
**Pull request by goaway**
Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

**Comment by goaway** [comment]
We could do it in native code, but it would unnecessarily add the overhead of the JNI call.

File: library/java/src/io/envoyproxy/envoymobile/engine/EnvoyHTTPStream.java [hunk 1]
```diff
@@ -38,7 +38,13 @@ public void sendHeaders(Map<String, List<String>> headers, boolean endStream) {
    * @param endStream, supplies whether this is the last data in the streamHandle.
    */
```
Comment (junr03) [inline library/java/src/io/envoyproxy/envoymobile/engine/EnvoyHTTPStream.java]: lets put a @throws here

File: library/java/src/io/envoyproxy/envoymobile/engine/JniLibrary.java [hunk 2]
```diff
@@ -69,6 +69,17 @@ public static void load() {
...
+  /**
+   * Send data over an open HTTP stream. This method can be invoked multiple
+   * times.
+   *
+   * @param stream,    the stream to send data over.
+   * @param data,      the data to send.
+   * @param endStream, supplies whether this is the last data in the stream.
+   * @return int, the resulting status of the operation.
+   */
+  protected static native int sendData(long stream, byte[] data, boolean endStream);
```
Comment (buildbreaker) [inline library/java/src/io/envoyproxy/envoymobile/engine/JniLibrary.java]: Should we update the other method's documentation to call out that we need to direct buffer?

**Comment by buildbreaker** [comment]
I think it's fine to do this check in the java layer

**Pull request merged**